specification/under_development/profiles/Base_Profile.md

```diff
@@ -0,0 +1,175 @@
+
+To indicate the total number of qubits required to execute the program, a custom attribute with the name `required_qubits` is defined and attached to the entry point. To indicate the number of registers/bits needed to store measurement results during program execution, a custom attribute with the name `required_results` is defined and attached to the entry point. The value of both of these attributes is the string representation of a 64-bit integer constant.
+
+TODO: why the string value, and not an integer?   
+TODO: no double underscore guard for attributes likely makes sense
+
+### Function Body
+
+The function body consists of a single block named `entry`. This implies in particular that no branching is permitted inside a function's body.
+In the `entry` block, any number of calls to QIS functions may be performed.
+To be compatible with the Base Profile these functions must return void. 
+Any arguments to invoke them must be inlined into the call itself; 
+they hence must be constants or a pointer to a [qubit or result]()
+The result of the program execution should be logged using output recording functions.
+
+The following instructions are the *only* LLVM instructions that are permitted within a Base Profile compliant program:
+
+| LLVM Instruction                  | Context and Purpose         | Rules for Usage |
+|---------------------------|-------------------|-------------|
+| `call`    | Used within a function block to invoke any one of the declared QIS functions and the output recording functions. | May optionally be preceded by a [`tail` marker](https://llvm.org/docs/LangRef.html#call-instruction). |
+| `ret` | Used to return the exit code of the program. | Must occur (only) as the final instruction of the `entry` block. |
+| `inttoptr` | Used to cast an `i64` integer value to either a `%Qubit*` or a `%Result*`. | May be used as part of a function call only. |
+| `getelementptr inbounds` | Used to create an `i8*` to pass a constant string for the purpose of labeling an output value. | May be used as part of call to an output recording function only. |
+
+## Data Types and Values
+
+Within the base profile, defining local variables is not supported. Arguments to calls correspondingly are expected to be constant values. 
+
+Qubits and results are passes as a pointer of type `%Qubit*` and `%Result*` respectively, where the pointer itself rather than the memory location identifies the qubit or result: For the purpose of passing them as arguments in function calls, a 64-bit signed integer value is cast to the appropriate pointer type. The signed integer that is case must be in the interval `[0, nrQubits)` for `%Qubit*` and `[0, nrResults)` for `%Result*`, where `nrQubits` and `nrResults` are the required number of qubits and results defined by entry point attributes.
```

Comment:
We've talked about the base profile being a proper subset of all extended profiles. Does this mean that for any future profiles (even those that directly support dynamic qubit allocation function calls that return %Result* values to a virtual register), this indexing format of casting integers to pointers needs to be supported?

Reply:
Casting int to point is a way to accurately reflect program intent for quantum programming frameworks that allow to express the same (i.e. frameworks where you e.g. have $0 to identify a qubit). The way it is intended to work is that it is a strict either-or choice; qubits are either encoding a qubit id and they are not pointing to a valid memory location, or they are actual pointers and the runtime manages them. As written, the Base Profile requires to use the former, while the intent for the full spec and other profiles is that either one may be chosen by the frontend. No longer supporting the const int to pointer I think is something to discuss when we have more experience with backends managing qubits.

Let me know what you think.